WireDomain/Sources/WireDomain/Notifications/GenerateNotificationService.swift

```diff
@@ -0,0 +1,170 @@
+    ) {
+        self.eventsStream = eventsStream
+        self.contentHandler = contentHandler
+        self.accountManager = accountManager
+        self.selectedAccount = selectedAccount
+        self.accountID = accountID
+        self.userDefaults = userDefaults
+        self.conversationLocalStore = conversationLocalStore
+        self.userLocalStore = userLocalStore
+        self.messageLocalStore = messageLocalStore
+    }
+
+    /// Processes the events stream.
+    func process() async {
+        for await events in eventsStream {
+            await generateNotifications(for: events)
+        }
+    }
+
+    private func generateNotifications(
+        for events: [UpdateEvent]
+    ) async {
+        guard !events.isEmpty else {
+            return contentHandler(.emptyNotification)
+        }
+
+        var notifications: [UNMutableNotificationContent] = []
+
+        for event in events {
+            var notificationBuilder: NotificationBuilder
+
+            switch event {
+            case let .conversation(conversationEvent):
+                notificationBuilder = await ConversationEventNotificationBuilder(
+                    event: conversationEvent,
+                    userID: accountID,
+                    userDefaults: userDefaults,
+                    userLocalStore: userLocalStore,
+                    conversationLocalStore: conversationLocalStore,
+                    messageLocalStore: messageLocalStore
+                )
+
+            case let .user(userEvent):
+                notificationBuilder = UserNotificationBuilder(
+                    event: userEvent,
+                    userLocalStore: userLocalStore
+                )
+
+            default:
+                continue
+            }
+
+            guard await notificationBuilder.shouldBuildNotification() else {
+                continue
+            }
+
+            do {
+                let userNotification = try await notificationBuilder.buildContent()
+
+                switch userNotification {
+                case let .text(notificationContent):
+                    notifications.append(notificationContent)
+                case let .callKit(callKitContent):
+                    do {
+                        try await CXProvider.reportNewIncomingVoIPPushPayload(callKitContent)
+                    } catch {
+                        WireLogger.calling.error(
+                            "failed to wake up main app: \(error.localizedDescription)",
+                            attributes: .newNSE
+                        )
+                    }
+                case .notDisplayed:
+                    continue
+                }
+
+            } catch {
+                WireLogger.notifications.error(
+                    "Failed to generate notification: \(error.localizedDescription)",
+                    attributes: .newNSE
+                )
+            }
+        }
+
+        await showNotifications(notifications)
```

Comment:
issue: I think also by calling this at the end of this method we consume the content handler too early, there may be more events left in the stream.